2단계 - 요구 사항 정리

NoJaeHyuk opened the pull request:
2단계 미션에 대한 PR입니다. 

.http 파일을 보면서 Controller -> Service 로 역추적하면서 요구사항을 정리하였습니다.

jhnoh7 committed:
docs: 키친포스에 대한 요구사항 정의

Hyeon9mak commented on README.md:
```diff
@@ -9,6 +9,95 @@ docker compose -p kitchenpos up -d
 
+- 식당이나 카페 등의 음식 서비스 업체에서 사용할 수 있는 주문 관리 시스템(키오스크)을 구현합니다.
+
+### 메뉴 관리
+- 메뉴그룹
+  - 식당이나 카페에서 제공하는 메뉴들의 카테코리를 분류하기 위한 단위입니다.
+  - [ ] 메뉴 그룹을 생성할 수 있다. 메뉴 그룹은 고유한 이름을 가지고 있어, 중복된 이름은 사용할 수 없다. 
+  - [ ] 모든 메뉴 그룹을 조회 할 수 있다.
+- 메뉴 
+  - [ ] 메뉴를 등록 할 수 있다.
+    - [ ] 메뉴는 메뉴 그룹안에 포함되어야한다.
+    - [ ] 메뉴의 이름, 가격, 노출여부, 구성 상품들을 필수로 기입해야한다. 
+      - [ ] 메뉴의 가격은 `0원`보다 적을 수 없다. 
+      - [ ] 메뉴의 이름은 비속어(profanity)가 들어갈 수 없다.
+      - [ ] 메뉴의 `구성 상품`은 이미 등록된 상품으로만 등록이 가능하다.
+      - [ ] 메뉴의 `구성 상품`수량은 0보다 작을 수 없다. 
+      - [ ] 메뉴의 가격이 `구성 상품`가격의 총합보다 클 수 없다. 
+  - [ ] 메뉴의 가격 수정이 가능하다. 
+    - [ ] 수정할 가격(메뉴의 가격)은 `0원`보다 작을 수 없다.
+    - [ ] 수정할 가격(메뉴의 가격)이 `구성 상품`가격의 총합보다 클 수 없다. 
```
개발자가 아니어도 읽고 충분히 이해할 수 있도록 친절히 요구사항을 정리해주셨네요 💯

Hyeon9mak commented on README.md:
```diff
@@ -9,6 +9,95 @@ docker compose -p kitchenpos up -d
+  - 메뉴를 비노출상태로 변경 할 수 있다. 
+  - 등록된 메뉴 목록을 조회 할 수 있다. 
+
+### 상품 
+- [ ] 상품을 등록 할 수 있다. 
+  - [ ] `이름`, `가격`을 필수로 입력하여야 한다.
+    - [ ] 가격은 `0원`보다 작을 수 없다.
+    - [ ] 이름은 비속어를 입력할 수 없다.
+- [ ] 상품의 가격을 변경할 수 있다. 
+  - [ ] 변경 가격은 `0원`보다 작을 수 없다. 
+  - [ ] 해당 상품으로 구성된 메뉴 가격이 수정된 상품의 가격 총합보다 커지게되면 해당 메뉴를 비노출시킨다. 
+- 등록한 상품 목록을 조회할 수 있다.
```
간단한 예시를 하나씩 포함시키는 건 어떨까요? 조금 귀찮긴 하지만, 예시들을 나열해두면 이해도도 높아지고
여러 사람이 함께 읽으므로 많은 사람들의 생산성에 큰 기여를 하게 될거에요!

NoJaeHyuk replied:
`
 - **상품 등록 예시**
    - **이름**: 후라이드 치킨
    - **가격**: 15,000원
`
`
예를 들어, `후라이드 치킨`의 상품 가격이 `19000원`에서 `24000원`으로 변경되면, `후라이드 치킨 세트`메뉴의 기존 가격이 `22000원`이라면 이 메뉴는 비노출처리가 됩니다.
`
와 같은 요구사항에 예시가 필요한 부분에 추가했습니다.

Hyeon9mak commented on README.md:
```diff
@@ -9,6 +9,95 @@ docker compose -p kitchenpos up -d
+    - [ ] 주문 시 테이블정보가 포함되어 있어야한다. (주문 전 테이블을 선정 받아야 한다.)
+    - [ ] 손님이 테이블 점유가 안됬을 시 주문이 불가하다.
+
+- [ ] 주문을 승인할 수 있다. 
+  - [ ] 주문의 상태는 `대기`, `접수승인`, `서빙완료`, `배송중`, `배송완료`, `완료`가 있다. 
+  - [ ] 주문의 상태가 `대기` 상태여야 한다.
+  - [ ] 주문유형이 `배달`일 경우 라이더에게 주문정보를 전달해야한다.
+    - [ ] `주문번호`, `가격`, `배송주소` 정보를 전달한다.
+  - [ ] 주문의 상태를 `접수승인`으로 변경한다.
+
+- [ ] 주문된 메뉴를 서빙할 수 있다.
+  - [ ] 주문의 상태가 `접수승인`여야 한다.
+  - [ ] 주문의 상태를 `서빙완료`로 변경한다.
+
+- [ ] 주문된 메뉴를 배송을 시작 할 수 있다.
+  - [ ] 주문 유형이 `배달`이어야 한다. 
+  - [ ] 주문된 메뉴가 완성(서빙완료)되어 있어야 한다. 
+  - [ ] 주문의 상태를 `배송중`으로 변경한다. 
+  
+- [ ] 주문된 메뉴의 배송 완료할 수 있다. 
+  - [ ] 주문의 상태가 `배송중`이어야한다.
+  - [ ] 주문의 상태를 `배송완료`로 변경한다. 
+
+- [ ] 주문을 완료처리 할 수 있다. 
+  - [ ] 주문의 유형이 `배달`인 경우 주문 상태가 `배송완료`여야 한다. 
+  - [ ] 주문의 유형이 `테이크아웃` 또는 `매장식사`인 경우 `서빙완료`여야 한다. 
+  - [ ] 주문의 상태를 `완료`로 변경한다.
+  - [ ] 주문의 유형이 `매장 식사`인 경우 사용 중인 완료 후 테이블을 정리하여야한다. 
+    - [ ] 테이블의 손님의 수를 `0`으로 변경한다. 
+    - [ ] 테이블의 점유여부를 `false`로 변경한다.
```
`주문 상태` 별로 계속해서 `주문 유형`에 대해 정의를 내려주고 있어요.
반대로 `주문 유형` 별로 나누어 `주문 상태`를 다뤄보는 건 어떨까요?

같은 주문처럼 보이지만, 사실 `매장식사`와 `배달`은 서로 완전히 다른 개념으로 볼 수도 있겠어요!

Hyeon9mak replied:
점유여부를 `false` 로 변경한다는 개발자 관점의 설명 같아요! 다른 설명은 없을까요? 😉

NoJaeHyuk replied:
주문 상태로 분류되면 아무래도 해당 주문 상태가 없는 유형에 대한 구분이나 유형별 주문하는 형태에서 한눈에 요구사항을 파악하기 힘들어 `주문 후 주문유형 별 주문처리 과정`을 만들어 재정의 하였습니다. 

`false`, `true` 같은 개발적 용어는 제거하고 사용중/미사용 으로 변경하였습니다.

Hyeon9mak commented on README.md:
```diff
@@ -9,6 +9,95 @@ docker compose -p kitchenpos up -d
+
+- [ ] 주문한 목록을 조회할 수 있다.
+
+### 주문-테이블(좌석)
+- [ ] 테이블을 등록할 수 있다. 
+  - [ ] 테이블의 이름을 부여할 수 있다. 
+  - [ ] 테이블 등록 시 `손님 0/빈자리`로 세팅이 된다.
+- [ ] 손님이 테이블을 점유할 수 있다. 
+  - [ ] 테이블 점유여부가 `true`가 된다.
+- [ ] 손님이 테이블을 사용 완료하면 테이블을 정리할 수 있다. 
+  - [ ] 테이블의 상태를 `손님0/빈자리`로 세팅할 수 있다.
+- [ ] 테이블 점유 시 손님의 수를 세팅할 수 있다. 
+- [ ] 테이블 목록을 조회할 수 있다.
+
```
테이블(좌석) 역시도 `매장식사`에만 포함되는 개념이죠 😄

Hyeon9mak replied:
점유, 정리, 세팅 등 용어가 혼용되고 있는 것 같아요!
용어를 모두 하나로 통일해주면 코드를 작성할 때도 고민하지 않고 큰 편의를 얻을 수 있겠는데요? 😄

NoJaeHyuk replied:
테이블 관리에 대한 부분에는 테이블 등록/조회/정리에 대한 부분만 정의하였고 
그 외 점유에 대한 부분은 `매장식사`유형 정리로 이동하였습니다. 

점유, 세팅은 테이블의 상태라는 표현으로 (손님 x명, 사용/미사용) 표기하는 식으로 공통적으로 변경하였습니다.
정리만 테이블을 치운다는 의미로 사용되게 변경하였습니다.

Hyeon9mak reviewed (changes requested):
안녕하세요 재혁님! 요구사항 잘 정리해주셨습니다 👍 
비개발직군도 쉽게 읽고 이해할 수 있도록 작성해주셨네요~
아주 조금이 코멘트만 달아두었으니 확인해주세요!

jhnoh7 committed:
docs: 키친포스에 대한 요구사항 업데이트

- 기존 주문 상태별 요구사항 정리를 주문유형별로 변경함.
- 테이블 관리 항목 일부를 `매장식사`에 정의함.
- 테이블 관리 관련 용어를 정리함. (점유,정리 세팅 등)
- 요구사항에 대한 필요시 예시를 추가함.

Hyeon9mak commented on README.md:
```diff
@@ -9,6 +9,122 @@ docker compose -p kitchenpos up -d
+      - [ ] 메뉴의 `구성 상품`은 이미 등록된 상품으로만 등록이 가능하다.
+      - [ ] 메뉴의 `구성 상품`수량은 0보다 작을 수 없다. 
+      - [ ] 메뉴의 가격이 `구성 상품`가격의 총합보다 클 수 없다.
+  - **메뉴 등록 예시**
+    - **메뉴그룹**: 추천메뉴
+    - **이름**: 후라이드 치킨세트
+    - **가격**: 22,000원
+    - **노출 여부**: 예
+    - **구성 상품**:
+      - 상품: 후라이드 치킨(가격: 20,000원)
+      - 수량: 1
+      - 상품: 콜라(가격: 3,000원)
+      - 수량: 1
```
👍 👍 재혁님은 조금 귀찮으셨겠지만, 재혁님 덕분에 많은 사람들이 도움을 받겠네요!!

Hyeon9mak commented on README.md:
```diff
@@ -9,6 +9,122 @@ docker compose -p kitchenpos up -d
+### 주문
+- [ ] 주문을 할 수 있다.
+  - [ ] 주문 유형을 입력해야 한다. 
+    - [ ] 주문 유형에는 `배달`, `테이크아웃`, `매장식사`가 있다. 
+  - [ ] 주문할 메뉴의 `정보(메뉴, 수량)`를 입력해야한다.
+    - [ ] 메뉴는 `노출`된 메뉴여야한 한다. 
+    - [ ] 주문한 메뉴와 실제 메뉴 가격이 일치해야 한다.
+    - [ ] `매장식사`를 제외하고 수량이 0 보다 적을 수 없다.
+- [ ] 주문 유형 `배달`로 주문이 가능하다. 
+  - [ ] 주문 시  배송주소를 반드시 입력해야한다.
+- [ ] 주문 유형이 `매장식사`로 주문이 가능하다. 
+  - [ ] 주문 시 테이블정보를 입력받아야 한다.
+    - [ ] 손님 또는 매장직원은 주문하기 전 테이블 번호와 인원 수를 입력하여야한다.
+    - 만약 테이블에 부착된 키오스크를 이용한다면 인원 수 만을 입력한다.
+    - [ ] 입력 시 테이블은 `손님 (수)명/사용중` 상태가 된다.
+  - [ ] 손님이 테이블의 사용상태가 아니라면 주문이 불가하다.
+- [ ] 주문한 목록들을 조회할 수 있다.
+
+### 주문 후 주문유형 별 주문처리 과정 
+- 주문의 상태는 `대기`, `접수승인`, `서빙완료`, `배송중`, `배송완료`, `완료`가 있다.
+- [ ] 주문이 온다면 초기 상태는 `대기`이다.
+
+- [ ] 주문 유형 `배달`로 주문 요청이 올 수 있다. 
+  - [ ] 주문을 `승인`할 경우 `라이더`에게 주문정보를 전달한다. 
+    - [ ] `주문번호`, `가격`, `배송주소` 정보를 전달한다.
+    - [ ] 전달되면 `접수승인` 상태가 된다. 
+  - [ ] 주문한 메뉴가 완성되어 라이더에게 전달한다. 
+    - [ ] `서빙완료`상태를 의미한다.
+  - [ ] 주문된 메뉴에 대해 배송을 시작할 수 있다. 
+    - [ ] 주문된 메뉴가 완성(서빙완료)되어 있어야 한다.
+    - [ ] 배송이 시작되면 주문의 상태가 `배송중`이 된다. 
+  - [ ] 주문된 메뉴의 배송 완료할 수 있다.
+    - [ ] 전 주문의 상태가 `배송중`이어야 `배송완료`를 할 수 있다.
+    - [ ] 주문의 상태를 `배송완료`로 변경한다.
+  - [ ] 매장은 주문된 메뉴에 대한 `완료`처리를 할 수 있다. 
+    - [ ] 전 주문의 상태가 `배송완료`여야 한다. 
+    - [ ] 주문의 상태가 `완료`로 변경된다.
+    
+- [ ] 주문 유형 `테이크아웃`으로 주문 요청이 올 수 있다.
+  - [ ] 주문을 승인 시 주문상태 `접수승인`이 된다.
```
👍 👍 다음 미션을 진행하실 때 재혁님께서 정의하신 요구사항에 따라 클래스 파일을 수정하는 경험을 하게 되실거에요!
(멤버변수 선언, public method 정의 등)
그 때 요것들이 어떻게 사용될지 기대되네요! 👍 👍

Hyeon9mak reviewed (approved):
안녕하세요 재혁님! 처음부터 요구사항 문서를 잘 정의해주신 덕분에 변경이 크지 않았네요! 👍 
마지막 단계로 넘어가시죠!!
